### PR TITLE
Use native tools to get pod ip for network plugins

### PR DESCRIPTION
**Pull request by dshulyak**
related #30144 #26093

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30225)

### CODE REVIEW

File: pkg/kubelet/util/podip/podip.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	"fmt"
+	"net"
+	"strings"
+
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/util/ioutils"
+)
+
+func getIpGetterCommand(interfaceName, addrType string) []string {
+	return []string{"ip", "-o", addrType, "addr", "show", "dev", interfaceName, "scope", "global"}
```
Comment (luxas) [inline pkg/kubelet/util/podip/podip.go]: This maybe should use `netlink.AddrList` like https://github.com/luxas/kubernetes/blob/5bdb377dcec9a2e9b42c81640186ae334bc6e46e/pkg/kubelet/container_bridge.go#L143 if possible.

I know it was like this before, but what about containers without the `ip` command?
(For example a `FROM scratch` container with a statically linked binary)

File: pkg/kubelet/util/podip/podip.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	if len(lines) < 1 {
+		return nil, fmt.Errorf("Unexpected command output %s", output)
+	}
+	fields := strings.Fields(lines[0])
+	if len(fields) < 4 {
+		return nil, fmt.Errorf("Unexpected address output %s ", lines[0])
+	}
+	ip, _, err := net.ParseCIDR(fields[3])
+	if err != nil {
+		return nil, fmt.Errorf("CNI failed to parse ip from output %s due to %v", output, err)
```
Comment (luxas) [inline pkg/kubelet/util/podip/podip.go]: CNI is not relevant here, maybe just `failed to parse ip from container exec output %s due to %v`

File: pkg/kubelet/network/cni/cni.go [hunk 3]
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/kubernetes/pkg/apis/componentconfig"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 	"k8s.io/kubernetes/pkg/kubelet/network"
+	"k8s.io/kubernetes/pkg/kubelet/podip"
```
Comment (luxas) [inline pkg/kubelet/network/cni/cni.go]: should be `k8s.io/kubernetes/pkg/kubelet/util/podip`

**Comment by luxas** [comment]
nice! but still I'd like to get rid of the `ip` dep

**Comment by luxas** [comment]
Hmm, after thinking a little bit more about it, I don't know if this is the right approach.
As I understand it; the current nsenter solution only joins the net interface and calls the ip binary that is located on host.

If we would use a native exec solution; all namespaces (not only net) would be joined, which makes it impossible to call ip on host => will fail for all containers without ip.

What do you think?

**Comment by dshulyak** [comment]
@luxas you are right, i completely missed this difference and thought that this is ok to assume that iproute2 will be present in infra container

**Comment by dshulyak** [comment]
Looks like there is 2 options:
1. Use vishvananda/netns library to enter container namespace and obtain ip address with netlink
2. Reuse CNI AddToNetwork call to get ip address (suggestion from code)

**Comment by luxas** [comment]
@dshulyak If you get 1 working, it would be great I think

cc @kubernetes/sig-network What about obtaining the IP address by using netlink and netns?

**Comment by bboreham** [comment]
> 1. Use vishvananda/netns library to enter container namespace

Note there is no safe way to do this in Go, except by exec-ing a separate process.  See https://github.com/vishvananda/netns/issues/17 and linked material for background.

**Comment by dshulyak** [comment]
It doesnt look like bridge cni plugin is idempotent 

```
Error adding container to network: failed to find a unique veth name
```

**Comment by thockin** [comment]
@freehan

**Comment by euank** [comment]
I'll throw in that the rkt runtime doesn't have an infra container (though it has a stage1) so adding the `ip` command to the infra container isn't a suitably cross-runtime option.

**Comment by freehan** [comment]
Close this for now. Please reopen when ready

**Pull request closed without merging**